src/main/java/uk/sky/cirrus/locking/CassandraLockingMechanism.java

```diff
@@ -0,0 +1,126 @@
+package uk.sky.cirrus.locking;
+
+import com.datastax.driver.core.PreparedStatement;
+import com.datastax.driver.core.ResultSet;
+import com.datastax.driver.core.Row;
+import com.datastax.driver.core.Session;
+import com.datastax.driver.core.exceptions.DriverException;
+import com.datastax.driver.core.exceptions.WriteTimeoutException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import uk.sky.cirrus.locking.exception.CannotAcquireLockException;
+import uk.sky.cirrus.locking.exception.CannotReleaseLockException;
+
+public class CassandraLockingMechanism extends LockingMechanism {
+
+    private static final Logger log = LoggerFactory.getLogger(CassandraLockingMechanism.class);
+
+    private final Session session;
+
+    private PreparedStatement insertLockQuery;
+    private PreparedStatement deleteLockQuery;
+    private boolean isRetryAfterWriteTimeout = false;
+
+    public CassandraLockingMechanism(Session session, String keyspace, CassandraLockConfig lockConfig) {
+        super(keyspace + ".schema_migration", lockConfig.getClientId());
+        this.session = session;
+    }
+
+    /**
+     * Prepares queries for acquiring and releasing lock.
+     *
+     * @throws CannotAcquireLockException if any DriverException thrown while executing queries.
+     */
+    @Override
+    public void init() throws CannotAcquireLockException {
+        super.init();
+
+        try {
+            insertLockQuery = session.prepare("INSERT INTO locks.locks (name, client) VALUES (?, ?) IF NOT EXISTS");
+            deleteLockQuery = session.prepare("DELETE FROM locks.locks WHERE name = ? IF client = ?");
+
+        } catch (DriverException e) {
+            throw new CannotAcquireLockException("Query to prepare locks queries failed", e);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     * <p>
+     * Returns true if successfully inserted lock.
+     * Returns false if current lock is owned by this client.
+     * Returns false if WriteTimeoutException thrown.
+     *
+     * @throws CannotAcquireLockException if any DriverException thrown while executing queries.
+     */
+    @Override
+    public boolean acquire() throws CannotAcquireLockException {
+        try {
+            ResultSet resultSet = session.execute(insertLockQuery.bind(lockName, clientId));
+            Row currentLock = resultSet.one();
+            if (currentLock.getBool("[applied]") || clientId.equals(currentLock.getString("client"))) {
```

Comment:
Do we really want re-entrant locks on the client Id level? Is there a reason for that?

Problems I can see
- schema update fails, node is restarted and can then reacquire the lock automatically. since system is an unknown state we probably don't want that to happen (we want manual intervention at this point).
- multiple threads can acquire a lock in the same application with the same client id, and then proceed to try applying schema concurrently